qualtran/bloqs/multi_control_multi_target_pauli.py

```diff
@@ -0,0 +1,76 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import cached_property
+from typing import Dict, Tuple
+
+import cirq
+import numpy as np
+from attrs import frozen
+from cirq_ft import MultiControlPauli as CirqMultiControlPauli
+
+from qualtran import Bloq, CompositeBloq, Register, Signature
+from qualtran.cirq_interop import CirqQuregT, decompose_from_cirq_op
+
+
+@frozen
+class MultiControlPauli(Bloq):
+    """Implements multi-control, single-target C^{n}P gate.
+
+    Implements $C^{n}P = (1 - |1^{n}><1^{n}|) I + |1^{n}><1^{n}| P^{n}$ using $n-1$
+    clean ancillas using a multi-controlled `AND` gate.
+
+    References:
+        [Constructing Large Controlled Nots]
+        (https://algassert.com/circuits/2015/06/05/Constructing-Large-Controlled-Nots.html)
+
+    Args:
+        cvs: Control values. Expect a tuple of tuples of control values, where
+            each tuple of control values should be the same length as the number
+            of bits in the corresponding control register. E.g. cvs = ((1, 1),
+            (1,), (1,1,1)) would mean we have 3 control registers of sizes (2,
+            1, 3). Currently we assume only control values of 1, a zero would
+            signify anti controls.
+        target_pauli: The name of the Pauli gate ("X", "Y", or "Z")
+    """
+
+    cvs: Tuple[Tuple[int, ...], ...]
```

Comment:
I think at this point it's worth spending the additional time to create a `ControlRegister` object and expect a `Sequence[ControlRegister]` instead of `cvs: Tuple[Tuple[int, ...], ...]`. 

Specifically, control values can be more than just `Tuple[int, ...]`. We've run into this before in Cirq and as a result, we ended up creating the control_values class that represents a combination of control values. 

https://github.com/quantumlib/Cirq/blob/64e1a7fda696af542fb32c4a4c664f1c5d51546f/cirq-core/cirq/ops/control_values.py#L25


For backwards compatibility, we couldn't get rid of the `Tuple[int, ...]` interface and it made the signature of `gate.controlled()` and `operation.controlled_by()` an absolute nightmare. 

Since we have the time and flexibility to do things right this time, I'd suggest let's create the right design to represent control values and then use that everywhere in classes that expect a `cvs` (like this one).

Reply:
Is the cirq control_values design something you would like to see again here?

Reply:
See #300 -- a controlregister class is on the roadmap. I don't think it should be blocking implementation of this bloq